### ISSUE TITLE
With lazy_import tensorflow cannot be imported

### ISSUE THREAD

**Issue by wookayin** [tensorflow/tensorflow#33316]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04 (and gLinux)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): pip binary
- TensorFlow version: 1.15.0.{rc1,rc2,rc3}
- Python version: python 3.6.8
- Installed using virtualenv? pip? conda?: anaconda
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the problem**

The following works correctly:

```bash 
$ python -c 'import tensorflow as tf; print(tf.__version__)'
1.15.0-rc3
```

In an interactive shell:

```python
import tensorflow as tf
```

I get the following error: `ValueError: operator __getitem__ cannot be overwritten again on class <class 'tensorflow.python.framework.ops.Tensor'>`.

**This happens when [`lazy_import`](https://github.com/mnmelo/lazy_import) is installed.** I think tensorflow's internal lazy_import has a conflict with some external library. TF 1.14 has no problem.

Though lazy_import is a third-party library, I think what it is doing is pretty non-hacky. Since this bug has appeared since 1.15.dev, we will need to have this fixed in the release version (given that 1.15 will be the last 1.x release).

Traceback:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow/__init__.py", line 99, in <module>
    from tensorflow_core import *
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/__init__.py", line 34, in <module>
    from tensorflow._api.v1 import autograph
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/_api/v1/autograph/__init__.py", line 22, in <module>
    from tensorflow._api.v1.autograph import experimental
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/_api/v1/autograph/experimental/__init__.py", line 10, in <module>
    from tensorflow.python.autograph.core.converter import Feature
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/__init__.py", line 35, in <module>
    from tensorflow.python.autograph import operators
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/operators/__init__.py", line 40, in <module>
    from tensorflow.python.autograph.operators.control_flow import for_stmt
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/operators/control_flow.py", line 65, in <module>
    from tensorflow.python.autograph.operators import py_builtins
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/operators/py_builtins.py", line 28, in <module>
    from tensorflow.python.autograph.utils import py_func
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/utils/__init__.py", line 21, in <module>
    from tensorflow.python.autograph.utils.context_managers import control_dependency_on_returns
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/autograph/utils/context_managers.py", line 24, in <module>
    from tensorflow.python.ops import tensor_array_ops
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/ops/tensor_array_ops.py", line 35, in <module>
    from tensorflow.python.ops import array_ops
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/ops/array_ops.py", line 1045, in <module>
    ops.Tensor._override_operator("__getitem__", _slice_helper)
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 745, in _override_operator
    _override_helper(Tensor, operator, func)
  File "/$HOME/.miniconda3/envs/ace-base/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 177, in _override_helper
    (operator, clazz_object))
ValueError: operator __getitem__ cannot be overwritten again on class <class 'tensorflow.python.framework.ops.Tensor'>.
```

**Comment by mihaimaruseac** [comment]
I cannot reproduce:

```
(gh_lazy_import) mihaimaruseac@ankh:/tmp/gh_lazy_import$ pip install lazy_import tensorflow==1.15.0rc3
...
Successfully installed absl-py-0.8.1 astor-0.8.0 gast-0.2.2 google-pasta-0.1.7 grpcio-1.24.1 h5py-2.10.0 keras-applications-1.0.8 keras-preprocessing-1.1.0 lazy-import-0.2.2 markdown-3.1.1 numpy-1.17.2 opt-einsum-3.1.0 protobuf-3.10.0 six-1.12.0 tensorboard-1.15.0 tensorflow-1.15.0rc3 tensorflow-estimator-1.15.1 termcolor-1.1.0 werkzeug-0.16.0 wrapt-1.11.2
(gh_lazy_import) mihaimaruseac@ankh:/tmp/gh_lazy_import$ python
Python 3.6.8 (default, Jan  3 2019, 03:42:36) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> 
```

**Comment by wookayin** [comment]
Oh, that's interesting. Let me try to figure out how to reproduce it consistently.

**Comment by wookayin** [comment]
@mihaimaruseac Did you by any chance try importing lazy_import?

```
import lazy_import
tf = lazy_import.lazy_module("tensorflow")
import tensorflow as tf
print(tf.__version__)
```

**Comment by mihaimaruseac** [comment]
Tried, in 3 different ways

```
(gh_lazy_import) mihaimaruseac@ankh:/tmp/gh_lazy_import$ python
Python 3.6.8 (default, Jan  3 2019, 03:42:36) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lazy_import
>>> import tensorflow as tf
>>> tf.__version__
'1.15.0-rc3'
>>> 
(gh_lazy_import) mihaimaruseac@ankh:/tmp/gh_lazy_import$ python
Python 3.6.8 (default, Jan  3 2019, 03:42:36) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lazy_import
>>> tf1 = lazy_import.lazy_module("tensorflow")
>>> import tensorflow as tf
>>> tf.__version__
'1.15.0-rc3'
>>> tf1.__version__
'1.15.0-rc3'
>>> 
(gh_lazy_import) mihaimaruseac@ankh:/tmp/gh_lazy_import$ python
Python 3.6.8 (default, Jan  3 2019, 03:42:36) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lazy_import
>>> tf1 = lazy_import.lazy_module("tensorflow")
>>> import tensorflow as tf
>>> tf1.__version__
'1.15.0-rc3'
>>> tf.__version__
'1.15.0-rc3'
>>> 
```

**Comment by wookayin** [comment]
I tired all of the code above and it is indeed working as expected (printing version without the problem). I found that it was actually from another lazy_import of **numpy**, specifically, when numpy is yet to be fully loaded:

```
import lazy_import
np = lazy_import.lazy_module("numpy")
import tensorflow as tf
# gives ValueError: operator __getitem__ cannot be overwritten again on class <class 'tensorflow.python.framework.ops.Tensor'>.
```

I can reproduce this on 3 different environments including glinux and macOS, both anaconda and brew python.

Not sure tensorflow is to blame or numpy is, but let me now look into why it happens.

**Comment by mihaimaruseac** [comment]
Confirmed that it reproduces.

Following https://github.com/tensorflow/tensorflow/issues/32159#issuecomment-542452574 I might get rid of `__getitem__` in one lazy import path. Let's see if that also solves this.

**Comment by mihaimaruseac** [comment]
Can you try again please? There has been some code changes on that code path.

**Comment by image357** [comment]
It's still a problem. I'm using tensorflow 2.2.0.
Somehow you cannot lazy import numpy when tensorflow is used:

```python
import lazy_import
np = lazy_import.lazy_module("numpy")
import tensorflow as tf
```

throws
```python
Traceback (most recent call last):
  File "test2.py", line 3, in <module>
    import tensorflow as tf
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/__init__.py", line 41, in <module>
    from tensorflow.python.tools import module_util as _module_util
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/python/__init__.py", line 64, in <module>
    from tensorflow.python.framework.framework_lib import *  # pylint: disable=redefined-builtin
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/python/framework/framework_lib.py", line 52, in <module>
    from tensorflow.python.framework.importer import import_graph_def
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/python/framework/importer.py", line 28, in <module>
    from tensorflow.python.framework import function
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/python/framework/function.py", line 36, in <module>
    from tensorflow.python.ops import array_ops
  File "/home/koepke/.local/share/virtualenvs/tensorflow-0PogEeTj/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py", line 1428, in <module>
    _NON_AUTOPACKABLE_TYPES = set(np.core.numerictypes.ScalarType)
AttributeError: module 'numpy.core' has no attribute 'numerictypes'
```

**Comment by mihaimaruseac** [comment]
I am unsure if this is a TF issue. The codepaths that caused me to initially think this has been an issue are gone in TF 2.2

Now that we dropped support for Python <3.5, we can try converting our code to use Python's lazy loaders instead of a custom one, maybe that would solve the issue.

Though I don't think I'll be able to work on this until at least mid of August

**Comment by mihaimaruseac** [comment]
We now dropped support for Python3.5 too, so we should be able to convert to Python's lazy loaders instead of using custom ones. PRs welcome

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by image357** [comment]
Still the same

```python3
Python 3.8.7 (default, Jan 20 2021, 00:00:00) 
[GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lazy_import
>>> np = lazy_import.lazy_module("numpy")
>>> import tensorflow as tf
2021-02-01 22:07:07.217289: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-02-01 22:07:07.217314: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/__init__.py", line 41, in <module>
    from tensorflow.python.tools import module_util as _module_util
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/__init__.py", line 46, in <module>
    from tensorflow.python import data
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/__init__.py", line 25, in <module>
    from tensorflow.python.data import experimental
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/experimental/__init__.py", line 96, in <module>
    from tensorflow.python.data.experimental import service
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/experimental/service/__init__.py", line 140, in <module>
    from tensorflow.python.data.experimental.ops.data_service_ops import distribute
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/experimental/ops/data_service_ops.py", line 25, in <module>
    from tensorflow.python.data.experimental.ops import compression_ops
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/experimental/ops/compression_ops.py", line 20, in <module>
    from tensorflow.python.data.util import structure
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/data/util/structure.py", line 33, in <module>
    from tensorflow.python.ops import tensor_array_ops
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/ops/tensor_array_ops.py", line 38, in <module>
    from tensorflow.python.ops import array_ops
  File "/home/koepke/.local/share/virtualenvs/test-RD8o3laV/lib64/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 1498, in <module>
    _NON_AUTOPACKABLE_TYPES = set(np.core.numerictypes.ScalarType)
AttributeError: module 'numpy.core' has no attribute 'numerictypes'
```

with latest tensorflow 2.4.1.

If you give my some hints on how to fix that I would do it. Just need point to start at.

**Comment by sanatmpa1** [comment]
@wookayin,

It is working fine in recent stable version of TF i.e `2.6.0` and I am able to import tensorflow with lazy_import. Please take a look at the [gist here](https://colab.research.google.com/gist/sanatmpa1/d122e3dcef78ea794a6c36cd99e3c5bb/33316.ipynb). Thanks!

**Comment by mirh** [comment]
Just for the records, I believe this was fixed with 3bec74195ec5a242d8783fdfc718d45dc7210c4e
